### PR TITLE
feat: list_with_delimiter for file-based object storage

### LINKED ISSUE #688: Implement list_with_delimiter for File object store
Need this so the server can load configuration and WAL data if using the local file system for storage.

### PR DESCRIPTION
**Pull request by carols10cents**
Fixes #688 and does a bit of refactoring because the error enum was getting out of hand.

### CODE REVIEW

**Review by pauldix — APPROVED** [review]
Looks good, nice cleanup and now things will work with the filesystem!

File: object_store/src/disk.rs [hunk 1]
```diff
@@ -151,17 +264,21 @@ mod tests {
...
     use futures::stream;
+    use tempfile::TempDir;
 
     #[tokio::test]
     async fn file_test() -> Result<()> {
         let root = TempDir::new()?;
         let integration = File::new(root.path());
 
         put_get_delete_list(&integration).await?;
+        list_with_delimiter(&integration).await?;
```
Comment (alamb) [inline object_store/src/disk.rs:280]: is it right to say that this adds coverage to the file based `list_with_delimiter` function? 👍

**Review by alamb — COMMENTED** [review]
Looks like a (very) nice cleanup to me

**Review by alamb — APPROVED** [review]

File: object_store/src/disk.rs (hunk 1 above)
Comment (carols10cents) [reply object_store/src/disk.rs:280]: Correct!

**Pull request merged**